src/syntax.jl

```diff
@@ -0,0 +1,185 @@
+    parsed = JuliaSyntax.parseall(JuliaSyntax.SyntaxNode, file)
+    return SyntaxNodeWrapper(parsed)
+end
+
+#####
+##### Do stuff with the parsed files
+#####
+
+using JuliaSyntax
+using JuliaSyntax: head
+
+function is_method(node)
+    k = kind(node)
+    # This is a function like function f() 1+1 end
+    k == K"function" && return true
+    if k == K"="
+        # 1-line function definitions like f() = abc...
+        # Show up as an `=` with the first argument being a call
+        # and the second argument being the function body.
+        # We handle those here.
+        # We do not handle anonymous functions like `f = () -> 1`
+        kids = JuliaSyntax.children(node)
+        length(kids) == 2 || return false
+        kind(kids[1]) == K"call" || return false
+        return true
+    end
+    return false
+end
+
+function get_function_name_if_not_qualified(node) # assumes `is_method(node) == true`
+    k = kind(node)
+
+    local f
+    # TODO- write this code correctly instead of using try-catch
+    try
+        if k == K"function"
+            f = node.children[1].children[1]
+        else
+            kids = JuliaSyntax.children(node)
+            f = kids[1].children[1]
+        end
+    catch
+        return nothing
+    end
+    if isnothing(f.children)
+        return string(f)
+    else
+        # Qualified name; assumed to not be in this package.
+        # Could in fact be from a submodule; this case is not handled right now.
+        return nothing
+    end
+end
+
+function get_imported_names(node)
+    @assert kind(node) == K"import"
+    return string.(first.(JuliaSyntax.children.(JuliaSyntax.children(JuliaSyntax.children(node)[1])[2:end])))
+end
+
+
+function count_interesting_things(tree::SyntaxNodeWrapper)
+    counts = Dict{String,Int}()
+    functions = Dict{String,Int}()
+    docstring_functions = Dict{String,Int}()
+    imported_names = Set{String}()
+    items = PostOrderDFS(tree)
+    foreach(items) do wrapper
+        node = wrapper.node
+        k = kind(node.raw)
+        if is_method(node)
+            key = "method"
+            function_name = get_function_name_if_not_qualified(node)
+            if function_name !== nothing
+                functions[function_name] = get(functions, function_name, 0) + 1
+            end
+            counts[key] = get(counts, key, 0) + 1
+        elseif k == K"struct"
+            # These we increment once
+            key = string(k)
+            counts[key] = get(counts, key, 0) + 1
+        elseif k == K"export"
+            # These we count by the number of their children, since that's the number of exports/packages
+            # being handled by that invocation of the keyword
+            key = string(k)
+            counts[key] = get(counts, key, 0) + length(JuliaSyntax.children(node))
+        elseif k == K"doc"
+            kids = JuliaSyntax.children(node)
+            # Is this ever not a string?
+            kind(kids[1].raw) == K"string" || return
```

Comment:
This must always be a string now: In JuliaSyntax 0.4 I made sure all strings are wrapped in `K"string"`, regardless of whether they're literals or contain interpolations,

Reply:
Ah ok, thanks for letting me know!

In general, I am not sure what to do with expected invariants in this kind of code. I think maybe I should copy over `@maybecatch` from https://github.com/beacon-biosignals/SlackThreads.jl/blob/74351c2863ec9a1cf22732873d4d2816aa9c140d/src/SlackThreads.jl#L27-L49 so we can get errors when testing and emit them as logs (maybe debug logs) at runtime, since we want to be able to run this over General and all kinds of things can happen.

Reply:
You could just use an `@assert`? This particular invariant shouldn't be able to be broken regardless of broken package code - it's the parser's job to ensure that's true.

But I agree catch-and-log is also generally appropriate near the top level of a tool which you'll run across all of General.  It's nice for the tool to continue even if a single package breaks an invariant you expected when writing the code.

Reply:
> it's the parser's job to ensure that's true

This wouldn't be true if we needed to analyze code after macro expansion of course.  That involves user code and all bets are off. But we don't do that yet :-)

Reply:
Will the parser maintain that invariant even if the source code is somehow invalid?

Reply:
(Because currently we analyze all `*.jl` files, even ones that are never included in package code, and may not actually be Julia source code…)

Reply:
Yes, by the definition of a doc block really: we wouldn't parse something as a `K"doc"` unless it has a string first followed by an expression on the same or next line.

Reply:
Ah ok, makes sense. I put it in an assert but wrapped it in a try-catch. Sounds like I can remove that wrapper. The outer wrapper will get it still if something does go wrong still (but we will give up on all static analysis in that case).

Reply:
This looks promising! What is the status? What is missing?
I would be interested in: Number of functions, and average and maximal lines of code per function...

Reply:
Thanks, those are helpful to know @ufechner7 ! I got stuck trying to figure out when a function is being extended from another package/base (e.g. a new method for `getindex`) vs being introduced in this package. It is tricky in the presence of submodules. My latest thought has been to skip all that, for now at least, and just count "methods" rather than "functions". However I moved to first land https://github.com/JuliaEcosystem/PackageAnalyzer.jl/pull/86 separately.